riscv-sbi.adoc

```diff
@@ -1682,3 +1682,106 @@ Low bits from `mvendorid`.
 Low bits is SBI implementation ID. The firmware specific SBI extensions are
 for SBI implementations. It provides firmware specific SBI functions which
 are defined in the external firmware specification.
+
+== Trusted Execution Environment SBI Extension Space (EID #0x544545 "TEE")
+The Trusted Execution Environment Unit Extension divides cpu Execution
+Environment into two parts: REE (Rich Execution Environment) and TEE
+(Trusted execution environment), As shown in the picture below <<fig_tee1>>.
+To enable some applications to perform security-related services,
+which are placed in TEE.
+
+[#fig_tee1]
+.SBI TEE extensions runtimes
+image::riscv-sbi-tee1.png[width=1007,height=464]
+
+Well, one option is REE runs linux, TEE runs optee-os.
+When starting a security service, REE needs to convey 8 parameters
```

Comment:
hi @liushiwei007 ，

According to RISCV SBI calling convention(https://github.com/riscv-non-isa/riscv-sbi-doc/blob/master/riscv-sbi.adoc#binary-encoding), a0-a7 is used as parameter between the supervisor and the SEE. a6 is funid, a7: extid.

just like RISCV SBI calling convention, ARM SMC calling convention define parameters passing during SMC exceptions.
(https://documentation-service.arm.com/static/6013e5faeee5236980d08619) 
ARM SMC argument passing is x0-x6(or r0-r6 on arch32), x0(or r0) always pass SMC FID(include module name and funid), result is returned in x0-x3.
Arm common parameter is r1-r6, which is same as riscv r0-r5.

So RISCV SBI calling convention a0-a7 is enough for OPTEE,only need to add a2,a3 to result list.

If a5 register is used as security state flag, riscv common data parameter is one less than arm, whether it is sufficient still need to be check. if TEE Extension in opensbi manage the secure state, there is no problem about parameters

thanks!

Reply:
hi,  @matthewgui ,
are you reading 2.6 SMC32/HVC32 argument passing? (https://documentation-service.arm.com/static/6013e5faeee5236980d08619) I see that the document says that R0-R7 is used to pass the argument.  
In this case,, one register is missing. The current design is to continue with the ARM-based command code and add the ext id "TEE". These arm commands are processed under the TEE handle, so two riscv registers are used. Unless arm commands are transcoded at the linux driver level or opensbi does not conflict with these commands, riscv's a7 registers are equivalent to arm's a0 register, in which case additional registers may not be used.
On the other hand, state flag management by opensbi is optional.

Reply:
hi  @liushiwei007 ,
yes, r7 is used for Hypervisor, if using hypervisor, maybe need to encode extid with funcid to a7.

thanks!

Reply:
> Unless arm commands are transcoded at the linux driver level or opensbi does not conflict with these commands, riscv's a7 registers are equivalent to arm's a0 register, in which case additional registers may not be used.

Hi @liushiwei007 
Do you have any idea about how RISCV's a7 is equivalent to ARM's a0 ?

Seems that each SBI extension has ID range from `extid_start` to `extid_end`.
Maybe we can define TEE extension with a ID range to be additional function ID which are mapping from ARM's commands a0 ?

```
struct sbi_ecall_extension ecall_optee = {
	.extid_start = SBI_EXT_OPTEE_START,
	.extid_end = SBI_EXT_OPTEE_END,
	.handle = sbi_ecall_tee_handler,
};
```

Reply:
The current design is the same for RISC-V and ARM code at the linux generic level, so the conversion is placed at the architectural level.  If pass the value of w0 directly to a7, it will definitely get an error;  If you want to convert the value, it is actually feasible, but I worry that the later optee-os function will be added or deleted will involve this part of the modification

Reply:
> The current design is the same for RISC-V and ARM code at the linux generic level, so the conversion is placed at the architectural level. If pass the value of w0 directly to a7, it will definitely get an error; If you want to convert the value, it is actually feasible, but I worry that the later optee-os function will be added or deleted will involve this part of the modification

Yes, SMCCC has additional bits such as fast call [30], service call [29:24], and function number [15:0].
Maybe we need to define RISC-V own calling convention for optee-os.